indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java

```diff
@@ -1271,6 +1278,138 @@ public void remove(final Task task)
     }
   }
 
+  /**
+   * Acquire a read lock to perform the segment transactional append action for a given datasource.
+   * Also verifies that all the locks are of the type APPEND for the task.
+   * @param task task to perform the append action
+   */
+  public void acquireTransactionalAppendLock(Task task)
+  {
+    acquireTransactionalAppendLock(task, LOCK_ACQUIRE_TIMEOUT_MILLIS);
+  }
+
+  @VisibleForTesting
+  void acquireTransactionalAppendLock(Task task, long lockAcquireTimeoutMillis)
+  {
+    for (TaskLockPosse lockPosse : findLockPossesForTask(task)) {
+      if (lockPosse.taskLock.getType() != TaskLockType.APPEND) {
+        throw new ISE(
+            "All the locks must be of type APPEND for segmentTransactionalAppend. Found lock of type[%s] for task[%s].",
+            lockPosse.taskLock.getType(), task.getId()
+        );
+      }
+    }
+    final String datasource = task.getDataSource();
+    giant.lock();
+    try {
+      final boolean acquired = datasourceToConcurrentLock
+          .computeIfAbsent(datasource, ds -> new ReentrantReadWriteLock())
+          .readLock()
+          .tryLock(lockAcquireTimeoutMillis, TimeUnit.MILLISECONDS);
```

Comment:
Doing this inside `giant.lock()` means that while one thread is waiting to acquire this lock, no other thread can acquire a lock even if it is for a different datasource/interval. That doesn't seem ideal.

Reply:
Would a concurrent map work? (Now that we have removed the other map)